agent/yubiagent/slot.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2022 Yahoo Inc.
+// Licensed under the terms of the Apache License 2.0. Please see LICENSE file in project root for terms.
+
+package yubiagent
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+	"github.com/theparanoids/crypki/proto"
+	"github.com/theparanoids/ysshra/attestation/yubiattest"
+	"github.com/theparanoids/ysshra/keyid"
+	keyutil "github.com/theparanoids/ysshra/sshutils/key"
+	"golang.org/x/crypto/ssh"
+)
+
+// SlotAgent wraps the yubiagent and stores the information of a key slot in the agent.
+type SlotAgent struct {
+	yubiAgent YubiAgent
+	code      string                                // the code of the key slot, e.g. "9a", "9e"
+	public    ssh.PublicKey                         // the public key stored in the key slot
+	attest    *x509.Certificate                     // the attestation certificate of the key slot
+	policy    keyid.TouchPolicy                     // the touch policy of the key slot
+	csrs      []*proto.SSHCertificateSigningRequest // the certificates signing requests that are backed by the key slot
+}
+
+// NewSlotAgent returns the key slot agent.
+func NewSlotAgent(yubiAgent YubiAgent, code string) (*SlotAgent, error) {
+	attestationCert, err := yubiAgent.AttestSlot(code)
+	if err != nil {
+		return nil, err
+	}
+	// Due to Infineon Technologies' RSA key generation issue, we do not support RSA
+	// certificates generated by YubiKeys with firmware between 4.2.6-4.3.4.
+	// Ref: https://www.yubico.com/support/security-advisories/ysa-2017-01/
+	if attestationCert.PublicKeyAlgorithm == x509.RSA && isAffectedVersion(attestationCert.Extensions) {
+		return nil, fmt.Errorf("found RSA certificate generated by YubiKey with firmware 4.2.6-4.3.4")
+	}
+
+	publicKey, err := ssh.NewPublicKey(attestationCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	if validateSSHPublicKeyAlgo(publicKey) {
+		return nil, fmt.Errorf("unsupported certificate type in the key slot: %s", code)
+	}
+
+	touchPolicy := getTouchPolicy(attestationCert)
+
+	return &SlotAgent{
+		code:      code,
+		public:    publicKey,
+		attest:    attestationCert,
+		policy:    touchPolicy,
+		yubiAgent: yubiAgent,
+	}, nil
+}
+
+// NewSlotAgentWithAttrs returns a new slot agent with attribute values.
+func NewSlotAgentWithAttrs(yubiAgent YubiAgent, code string, public ssh.PublicKey,
+	attest *x509.Certificate, policy keyid.TouchPolicy) *SlotAgent {
+	return &SlotAgent{
+		yubiAgent: yubiAgent,
+		code:      code,
+		public:    public,
+		attest:    attest,
+		policy:    policy,
+	}
+}
+
+// RegisterCSR registers the given certificate signing request for the key slot.
+func (s *SlotAgent) RegisterCSR(csr *proto.SSHCertificateSigningRequest) {
```

Comment:
I'm not quite sure why we need to keep the `CSR`. Can you provide use cases?

Reply:
We need both `CSR` and `AddCertsToAgent` in `struct slot`, because the struct implements the interface `csr.AgentKey`.  In this way, gensign main logic could extract all the CSRs for a key pair (or a key slot), sign them, and add the signed certificates to the corresponding agent.
```
	for _, agentKey := range csrAgentKeys {
		var (
			certs    []ssh.PublicKey
			comments []string
		)
		for _, csr := range agentKey.CSRs() {
			cert, comment, err := signer.Sign(ctx, csr)
			if err != nil {
				return fmt.Errorf("failed to sign CSR: %v", err)
			}
			certs = append(certs, cert...)
			comments = append(comments, comment...)
		}
		err = agentKey.AddCertsToAgent(certs, comments)
		if err != nil {
			return fmt.Errorf("failed to add certificates into the agent: %v", err)
		}
	}
```

Reply:
To me, it feels like we probably need to review the design. Now, it is more like we need those methods exported because the implementation requires this to do so, not the methods is available to the struct or object.
For more specific, an Yubikey PIV slot can contain private key, and certificate but not CSRs. It doesn't make sense the retrieve it from the agent just because the implementation needs it. Similar, the ssh-agent doesn't hold the CSRs.
Also, allow a slot to retrieve its agent might be helpful in implementation, but doesn't make sense to expose it.

Reply:
Checking the [Handler interface](https://github.com/theparanoids/ysshra/blob/f983bc60f133789d02defa6218e65bcc2360129c/gensign/handler.go#L18-L22)
It probably will make more sense to just pass the `csr.Signer` into the handler to sign the certificate and store into the agent.

Reply:
Per discussion, the new interface will be like:
```
type Handler interface {
	Name() string
	Authenticate(params *csr.ReqParam) error
	Sign(signer, param) error // 1. Gen CSR, 2. Sign, 3. Add into agent
}

```